### ISSUE TITLE
Constraint violation when creating MultiInventoryItems with same Product

### ISSUE THREAD

**Issue by martinmo** [st-tu-dresden/salespoint#283]
`MultiInventoryItems` should support saving the same product multiple times in a `MultiInventory`, but fails with an exception caused by a `JdbcSQLIntegrityConstraintViolationException`:

```
Caused by: org.h2.jdbc.JdbcSQLIntegrityConstraintViolationException: Eindeutiger Index oder Primärschlüssel verletzt: "PUBLIC.UK_ASXDOVFBIT98196UXDBU7QOBY_INDEX_7 ON PUBLIC.INVENTORY_ITEM(PRODUCT_PRODUCT_ID) VALUES 1"
Unique index or primary key violation: "PUBLIC.UK_ASXDOVFBIT98196UXDBU7QOBY_INDEX_7 ON PUBLIC.INVENTORY_ITEM(PRODUCT_PRODUCT_ID) VALUES 1"; SQL statement:
insert into inventory_item (quantity_amount, quantity_metric, product_product_id, dtype, item_id) values (?, ?, ?, 'MultiInventoryItem', ?) [23505-199]
```

[[Full strack trace: stacktrace.log](https://github.com/st-tu-dresden/salespoint/files/3887693/stacktrace.log)].

The error can be reproduced with the following `DataInitializer`:

```java
package kickstart.example;

import org.salespointframework.catalog.*;
import org.salespointframework.core.*;
import org.salespointframework.inventory.*;
import org.salespointframework.quantity.Quantity;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
public class MultiItemInitializer implements DataInitializer {
	@Autowired MultiInventory<MultiInventoryItem> inventory;
	@Autowired Catalog<Product> catalog;

	@Override
	public void initialize() {
		var product = catalog.save(new Product("Beispiel", Currencies.ZERO_EURO));
		inventory.save(new MultiInventoryItem(product, Quantity.of(1)));
		inventory.save(new MultiInventoryItem(product, Quantity.of(1)));
	}
}

```

The issue was originally reported by user *Lennoxon* on Auditorium, see <https://auditorium.inf.tu-dresden.de/en/questions/4257>.

**Comment by martinmo** [comment]
Investigation has shown that both `MultiInventoryItem` and `UniqueInventoryItem` are mapped to a single table that has a `UNIQUE` constraint on the column that is mapped to `product`. These are the DDL statements that Hibernate issues for them:
```
$ ./mvnw spring-boot:run | grep -E '(alter|create) table inventory_item'
Hibernate: create table inventory_item (dtype varchar(31) not null, item_id varchar(255) not null, quantity_amount decimal(19,2), quantity_metric integer, product_product_id varchar(255), primary key (item_id))
Hibernate: alter table inventory_item add constraint UK_asxdovfbit98196uxdbu7qoby unique (product_product_id)
Hibernate: alter table inventory_item add constraint FK1vea7e9dxae889gd8i0u6q1a1 foreign key (product_product_id) references product
```

**Comment by martinmo** [comment]
That means, the underlying issue is that both are mapped to a *single* table. We therefore should move the base class `Inventory` from `@Entity` to `@MappedSuperclass`.

Unfortunately, this issue wasn't caught in our test suite :( Most certainly because of a difference in the importing order of entity classes (which can be traced by setting the log level for `org.hibernate.cfg` to `DEBUG`). In the Salespoint test suite, entity `MultiInventoryItem` is loaded *before* `UniqueInventoryItem` and the column definition of the "multi product" wins. In kickstart projects, it happens to be the other way around. Generally, as @odrotbohm pointed out, the order of classes can not be relied upon.

**Comment by odrotbohm** [comment]
That's merged and backported into 7.2.x.
